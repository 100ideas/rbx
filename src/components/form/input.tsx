import { cx } from "emotion";
import React from "react";

import { Element } from "@/components/element";
import { ModifierProps } from "@/modifiers";
import { Colors } from "@/modifiers/colors";

export type InputModifierProps = Partial<{
  className: string;
  color: Colors;
  disabled: boolean;
  isStatic: boolean;
  /**
   * The name of the input field Commonly used for [multi-input handling](https://reactjs.org/docs/forms.html#handling-multiple-inputs)
   */
  name: string;
  placeholder: string;
  readOnly: boolean;
  size: "small" | "medium" | "large";
  style: React.CSSProperties;
  type:
    | "text"
    | "email"
    | "tel"
    | "password"
    | "number"
    | "search"
    | "color"
    | "date"
    | "time";
  value: string;
}>;

export type InputProps = ModifierProps &
  InputModifierProps &
  Partial<
    Omit<
      React.ComponentPropsWithoutRef<"input">,
      "color" | "size" | "unselectable"
    >
  >;

type InputControllerProps = InputProps & {
  innerRef: React.Ref<HTMLInputElement>;
};

class InputController extends React.PureComponent<InputControllerProps> {
  public static defaultProps = {
    disabled: false,
    isStatic: false,
    readOnly: false,
    type: "text",
  };

  public render() {
    const {
      className,
      color,
      disabled,
      isStatic,
      name,
      placeholder,
      readOnly,
      size,
      type,
      value,
      innerRef,
      ...props
    } = this.props;
    return (
      <Element
        {...props}
        ref={innerRef}
        renderAs="input"
        name={name}
        value={value}
        type={type}
        placeholder={placeholder}
        readOnly={readOnly || isStatic}
        disabled={disabled}
        className={cx("input", className, {
          "is-static": isStatic,
          [`is-${size}`]: size,
          [`is-${color}`]: color,
        })}
      />
    );
  }
}

export const Input = React.forwardRef<HTMLInputElement, InputProps>(
  (props, ref) => <InputController innerRef={ref} {...props} />,
);
