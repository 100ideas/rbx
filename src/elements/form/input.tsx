import classNames from "classnames";
import PropTypes from "prop-types";
import React from "react";

import { forwardRefAs, Generic, HelpersProps } from "../../base";
import { Colors, COLORS } from "../../base/helpers";
import { tuple } from "../../utils";

export const INPUT_SIZES = tuple("small", "medium", "large");
export type InputSizes = (typeof INPUT_SIZES)[number];

export const INPUT_STATES = tuple("focused", "hovered");
export type InputStates = (typeof INPUT_STATES)[number];

export const INPUT_TYPES = tuple(
  "text",
  "email",
  "tel",
  "password",
  "number",
  "search",
  "color",
  "date",
  "time",
);
export type InputTypes = (typeof INPUT_TYPES)[number];

export type InputModifierProps = Partial<{
  color: Colors;
  readOnly: React.InputHTMLAttributes<HTMLInputElement>["readOnly"];
  rounded: boolean;
  size: InputSizes;
  state: InputStates;
  static: boolean;
  type: InputTypes;
}>;

export type InputProps = HelpersProps & InputModifierProps;

const propTypes = {
  color: PropTypes.oneOf(COLORS),
  readOnly: PropTypes.bool,
  rounded: PropTypes.bool,
  size: PropTypes.oneOf(INPUT_SIZES),
  state: PropTypes.oneOf(INPUT_STATES),
  static: PropTypes.bool,
  type: PropTypes.oneOf(INPUT_TYPES),
};

export const Input = Object.assign(
  forwardRefAs<InputProps, "input">(
    (
      {
        className,
        color,
        readOnly,
        rounded,
        size,
        state,
        static: isStatic,
        ...rest
      },
      ref,
    ) => (
      <Generic
        className={classNames(
          "input",
          {
            [`is-${color}`]: color,
            "is-rounded": rounded,
            [`is-${size}`]: size,
            "is-static": isStatic,
            [`is-${state}`]: state,
          },
          className,
        )}
        readOnly={readOnly || isStatic}
        ref={ref}
        {...rest}
      />
    ),
    { as: "input" },
  ),
  { propTypes },
);
