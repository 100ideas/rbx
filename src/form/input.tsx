import { cx } from "emotion";
import React from "react";

import { Generic } from "@/generic";
import { ModifierProps } from "@/modifiers";
import { Colors } from "@/modifiers/color";
import { tuple } from "@/utils";

export const INPUT_SIZES = tuple("small", "medium", "large");
export type InputSizes = (typeof INPUT_SIZES)[number];

export const INPUT_STATES = tuple("focused", "hovered");
export type InputStates = (typeof INPUT_STATES)[number];

export const INPUT_TYPES = tuple(
  "text",
  "email",
  "tel",
  "password",
  "number",
  "search",
  "color",
  "date",
  "time",
);
export type InputTypes = (typeof INPUT_TYPES)[number];

export type InputModifierProps = Partial<{
  color: Colors;
  disabled: boolean;
  /**
   * The name of the input field Commonly used for [multi-input handling](https://reactjs.org/docs/forms.html#handling-multiple-inputs)
   */
  name: string;
  placeholder: string;
  readOnly: boolean;
  rounded: boolean;
  size: InputSizes;
  state: InputStates;
  static: boolean;
  type: InputTypes;
  value: string;
}>;

export type InputProps = Prefer<
  ModifierProps & InputModifierProps,
  React.HTMLAttributes<HTMLInputElement>
>;

type InputControllerProps = InputProps & {
  innerRef: React.Ref<HTMLInputElement>;
};

class InputController extends React.PureComponent<InputControllerProps> {
  public static defaultProps = {
    disabled: false,
    readOnly: false,
    static: false,
    type: "text",
  };

  public render() {
    const {
      className,
      color,
      disabled,
      innerRef,
      name,
      placeholder,
      readOnly,
      rounded,
      size,
      state,
      static: isStatic,
      type,
      value,
      ...props
    } = this.props;
    return (
      <Generic<"input">
        {...props}
        ref={innerRef}
        as="input"
        name={name}
        value={value}
        type={type}
        placeholder={placeholder}
        readOnly={readOnly || isStatic}
        disabled={disabled}
        className={cx("input", className, {
          [`is-${color}`]: color,
          "is-rounded": rounded,
          [`is-${size}`]: size,
          "is-static": isStatic,
          [`is-${state}`]: state,
        })}
      />
    );
  }
}

export const Input = React.forwardRef<HTMLInputElement, InputProps>(
  (props, ref) => <InputController innerRef={ref} {...props} />,
);
