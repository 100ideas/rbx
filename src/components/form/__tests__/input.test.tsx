import React from "react";
import renderer from "react-test-renderer";

import { Input } from "../input";

describe("Input component", () => {
  it("should exist", () => {
    expect(Input).toMatchSnapshot();
  });

  it("should have input classname", () => {
    const component = renderer.create(<Input />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("should concat classname in props with Bulma classname", () => {
    const component = renderer.create(
      <Input className="other-class this-is-a-test" />,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("should use inline styles", () => {
    const component = renderer.create(<Input style={{ height: 250 }} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("should be type email and a with success colors", () => {
    const component = renderer.create(<Input color="success" type="email" />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("should be large and readOnly", () => {
    const component = renderer.create(<Input readOnly size="large" />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("should be disabled with placeholder and value", () => {
    const component = renderer.create(
      <Input value="TEST" disabled placeholder="hello tests" />,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});
