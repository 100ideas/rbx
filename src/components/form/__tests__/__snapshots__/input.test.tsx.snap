// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component Should be disabled with placeholder and value 1`] = `
<input
  className="input"
  disabled={true}
  placeholder="hello tests"
  readOnly={false}
  type="text"
  value="TEST"
/>
`;

exports[`Input component Should be large and readOnly 1`] = `
<input
  className="input is-large"
  disabled={false}
  readOnly={true}
  type="text"
/>
`;

exports[`Input component Should be type email and a with success colors 1`] = `
<input
  className="input is-success"
  disabled={false}
  readOnly={false}
  type="email"
/>
`;

exports[`Input component Should concat classname in props with Bulma classname 1`] = `
<input
  className="input other-class this-is-a-test"
  disabled={false}
  readOnly={false}
  type="text"
/>
`;

exports[`Input component Should exist 1`] = `
Object {
  "$$typeof": Symbol(react.forward_ref),
  "render": [Function],
}
`;

exports[`Input component Should have input classname 1`] = `
<input
  className="input"
  disabled={false}
  readOnly={false}
  type="text"
/>
`;

exports[`Input component Should use inline styles 1`] = `
<input
  className="input"
  disabled={false}
  readOnly={false}
  style={
    Object {
      "height": 250,
    }
  }
  type="text"
/>
`;
